### PR TITLE
Onboarding Kafka benchmark workload [Windows]

### PR DESCRIPTION
**Pull request by mohitgole**
Added kafka workload support for windows x64 and arm64.
1. Added server/client executors.
2. Added producer/consumer parsers.
3. Added unit tests.
4. Added functional tests.
5. Added documentation.

Note: **Support for linux OS will be raised post current windows PR**

### CODE REVIEW

**Comment by mohitgole** [comment]
Document is in progress.

**Comment by mohitgole** [comment]
@microsoft-github-policy-service agree

File: src/VirtualClient/TestResources/Results_KafkaProducer.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+150295 records sent, 24406.5 records/sec (2.33 MB/sec), 85.4 ms avg latency, 1253.0 ms max latency.
```
Comment (yangpanMS) [inline src/VirtualClient/TestResources/Results_KafkaProducer.txt:1]: Just curious why the mb/sec keeps getting higher, is there a warming up we need to do? We need to make sure that the total runtime doesn't change the result

File: src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json [hunk 2]
```diff
@@ -0,0 +1,221 @@
+{
+  "Description": "Kafka Benchmark",
+  "Metadata": {
+    "RecommendedMinimumExecutionTime": "02:00:00",
+    "SupportedPlatforms": "linux-x64,win-x64",
```
[the remark this answers is not in the document]
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json]: Yes. Updated the profile

File: src/VirtualClient/TestResources/Results_KafkaProducer.txt (hunk 1 above)
Comment (mohitgole) [reply src/VirtualClient/TestResources/Results_KafkaProducer.txt:1]: This is generated by Kafka load generating tools shipped by kafka.

Comment (yangpanMS) [reply src/VirtualClient/TestResources/Results_KafkaProducer.txt:1]: No I understand that, what I'm saying is we need to make sure that if we need a warm up period

Comment (mohitgole) [reply src/VirtualClient/TestResources/Results_KafkaProducer.txt:1]: No, it is not like it always increases. I have added a producer file from one the instance run.
[2024-02-01t1013356098z-producer-numofrec-5000000-recordsize-100-batchsize-8196-async-3.log](https://github.com/microsoft/VirtualClient/files/14235651/2024-02-01t1013356098z-producer-numofrec-5000000-recordsize-100-batchsize-8196-async-3.log)

Also, warmup that I have used in the code is just to make sure that we produce data once and then keep on consuming it. If we try to produce it every time the memory required will be a lot higher.

**Comment by yangpanMS** [comment]
@mohitgole would you please schedule a meeting with me to review this PR? We might also opt to put this in our internal repo first and then move to open source

File: src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        "Port": "$.Parameters.ServerPort",
+        "ServerInstances": "$.Parameters.ServerInstances",
+        "NumberOfRecords": "$.Parameters.NumberOfRecords",
+        "CommandLine": "--topic sync-test-rep-one --messages {NumberOfRecords} --bootstrap-server {0}:{Port}",
+        "Role": "Client",
+        "CommandType": "ConsumerTest"
+      }
+    },
+    {
+      "Type": "ParallelExecution",
```
Comment (mohitgole) [inline src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json:94]: Need to change from Parallel Execution to running multiple processes within client executor.

Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json]: PR is updated where we no longer use ParallelExecution, instead we will use ClientInstances as a parameter.

File: src/VirtualClient/TestResources/Results_KafkaProducer.txt (hunk 1 above)
Comment (mohitgole) [reply src/VirtualClient/TestResources/Results_KafkaProducer.txt:1]: We will not log producer metrics as it will be executed once, then the consumer will keep on consuming the data and keep logging the metrics to kusto.

File: src/VirtualClient/VirtualClient.Main/VirtualClient.Main.csproj [hunk 4]
```diff
@@ -62,18 +62,18 @@
...
 
-    <!--
+	<!--
     Copy the script/content files into the target runtime publish directory.
     -->
-    <Target Name="CopyInScriptsToPublishDirectory" AfterTargets="Publish">
-        <ItemGroup>
-            <ScriptFiles Include="$(RepoPath)\out\scripts\**\*.*" />
-        </ItemGroup>
+	<Target Name="CopyInScriptsToPublishDirectory" AfterTargets="Publish">
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Main/VirtualClient.Main.csproj:68]: Please confirm your tab is 4 spaces.

File: src/VirtualClient/VirtualClient.Main/CommandBase.cs [hunk 5]
```diff
@@ -292,6 +292,11 @@ protected virtual IServiceCollection InitializeDependencies(string[] args)
                 this.LoggingLevel = LogLevel.Trace;
             }
 
+            if (this.Debug)
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Main/CommandBase.cs:295]: We already have this in, why do you need to add it?

File: src/VirtualClient/VirtualClient.Core/VirtualClientComponentExtensions.cs [hunk 6]
```diff
@@ -108,7 +113,9 @@ public static class VirtualClientComponentExtensions
             CancellationToken cancellationToken,
             bool runElevated = false,
             string username = null,
-            Action<IProcessProxy> beforeExecution = null)
+            Action<IProcessProxy> beforeExecution = null,
+            TimeSpan? timeout = null,
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Core/VirtualClientComponentExtensions.cs:117]: Please let me know why this change was needed.

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaExecutor.cs [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    using VirtualClient.Common.Platform;
+    using VirtualClient.Common.Telemetry;
+    using VirtualClient.Contracts;
+
+    /// <summary>
+    /// The Kafka client action
+    /// </summary>
+    [UnixCompatible]
+    [WindowsCompatible]
+    public class KafkaExecutor : VirtualClientComponent
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/Kafka/KafkaExecutor.cs:26]: make it abstract if not intended for direct use and call it like KafkaBaseExecutor

File: src/VirtualClient/VirtualClient.Actions.UnitTests/Kafka/KafkaProducerMetricsParserTests.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+            Assert.IsTrue(metrics.Where(m => m.Name.EndsWith("Records_Per_Sec")).All(m => m.Relativity == MetricRelativity.HigherIsBetter));
+            Assert.IsTrue(metrics.Where(m => m.Name.Contains("Latency-Avg")).All(m => m.Relativity == MetricRelativity.LowerIsBetter));
+        }
+
+        [Test]
+        public void KafkaProducerMetricsParserThrowIfInvalidOutputFormat()
+        {
+            string workingDirectory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            string outputPath = Path.Combine(workingDirectory, @"Examples\Kafka\KafkaProducerResultExample.txt");
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions.UnitTests/Kafka/KafkaProducerMetricsParserTests.cs:65]: Use platform agnostic Path.Combine so that this works in linux unit test

File: src/VirtualClient/VirtualClient.Main/CommandBase.cs (hunk 5 above)
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Main/CommandBase.cs:295]: Yeah this looks like something came from merge conflict. Will remove it.

File: src/VirtualClient/VirtualClient.Core/VirtualClientComponentExtensions.cs (hunk 6 above)
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Core/VirtualClientComponentExtensions.cs:117]: As discussed, creating similar method in KafkaExecutor and removing change in VirtualClientComponentExtensions.

File: src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      }
+    }
+  ],
+  "Dependencies": [
+    {
+      "Type": "DependencyPackageInstallation",
+      "Parameters": {
+        "Scenario": "DownloadKafkaPackage",
+        "BlobContainer": "packages",
+        "BlobName": "kafka_2.13-3.6.1.zip",
```
[the remark this answers is not in the document]
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Main/profiles/PERF-KAFKA.json:159]: Yeah, so here 2.13 is scala version and 3.6.1 is kafka version.

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaExecutor.cs (hunk 7 above)
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Actions/Kafka/KafkaExecutor.cs:26]: Yeah.

File: src/VirtualClient/VirtualClient.Actions.UnitTests/Kafka/KafkaProducerMetricsParserTests.cs (hunk 8 above)
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Actions.UnitTests/Kafka/KafkaProducerMetricsParserTests.cs]: Yes made the change.

File: website/docs/overview/overview.md [hunk 10]
```diff
@@ -91,6 +91,7 @@ for using them.
 | [Stressapptest](https://microsoft.github.io/VirtualClient/docs/workloads/stressapptest) | Fault Tolerance | linux-x64, linux-arm64 | [Apache-2.0](https://github.com/stressapptest/stressapptest/blob/master/NOTICE)  |
 | [Stress-ng](https://microsoft.github.io/VirtualClient/docs/workloads/stress-ng) | Fault Tolerance | linux-x64, linux-arm64 | [GPL-2.0](https://github.com/ColinIanKing/stress-ng/blob/master/COPYING)  |
 | [SuperBench](https://microsoft.github.io/VirtualClient/docs/workloads/superbenchmark) | Machine Learning | linux-x64 | [MIT](https://github.com/microsoft/superbenchmark/blob/main/LICENSE)  |
+| [Kafka](../workloads/kafka/kafka.md) | Streaming data | win-x64, win-arm64 | [Apache License 2.0](https://github.com/apache/kafka/blob/trunk/LICENSE)  |
```
Comment (yangpanMS) [inline website/docs/overview/overview.md:94]: Just Apache-2.0 to align with the others

File: website/docs/workloads/kafka/kafka.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+)
+```
+
+to
+```bat
+rem Classpath addition for release
+call :concat "%BASE_DIR%\libs\*;"
+```
+
+For reference [The input line is too long](https://github.com/kafka-dev/kafka/issues/61)
```
Comment (yangpanMS) [inline website/docs/workloads/kafka/kafka.md:31]: Thank you for documenting this

File: website/docs/workloads/kafka/kafka-profiles.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+* [Environment Layouts](../../guides/0020-client-server.md)
+
+In the environment layout file provided to the Virtual Client, define the role of the client system/VM as "Client" and the role of the server system(s)/VM(s) as "Server".
+The spelling of the roles must be exact. The IP addresses of the systems/VMs must be correct as well. The following example illustrates the
+idea. The name of the client must match the name of the system or the value of the agent ID passed in on the command line.
+
+``` bash
+# Single System (environment layout not required)
+./VirtualClient --profile=PERF-KAFKA.json --system=Juno --timeout=1440
```
Comment (yangpanMS) [inline website/docs/workloads/kafka/kafka-profiles.md:23]: Please remove all Juno references. If you find it elsewhere please remove too.

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs [hunk 13]
```diff
@@ -0,0 +1,498 @@
+﻿using System;
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs:1]: Add mit header please

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+            this.fileSystem = this.systemManagement.FileSystem;
+            this.serverProcesses = new List<Task>();
+            this.ServerRetryPolicy = Policy.Handle<Exception>(exc => !(exc is OperationCanceledException))
+                .WaitAndRetryAsync(10, (retries) => TimeSpan.FromSeconds(retries));
+        }
+
+        /// <summary>
+        /// Port on which server runs.
+        /// </summary>
+        public int Port
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs:52]: For workloads like these we can have a default port so profile can be shorter.

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+                        telemetryContext.Clone().AddError(exc));
+
+                    throw;
+                }
+            });
+        }
+
+        private async Task ConfigurePropertiesFileAsync(EventContext telemetryContext, CancellationToken cancellationToken)
+        {
+            string logDir = "/tmp";
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs:358]: Does this PR have Linux support or not? Just to make sure we are coding for both OS in the future

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaExecutor.cs [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+                this.ThrowIfRoleNotSupported(clientInstance.Role);
+            }
+
+            this.InitializeApiClients();
+
+            DependencyPath kafkaPackage = await this.GetPackageAsync(this.PackageName, cancellationToken);
+            this.KafkaPackagePath = kafkaPackage.Path;
+
+            switch (this.Platform)
+            {
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/Kafka/KafkaExecutor.cs:155]: I don't believe you would need this. You don't need cmd to act as a process exe

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaClientExecutor.cs [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Polly;
+using VirtualClient.Common;
+using VirtualClient.Common.Extensions;
+using VirtualClient.Common.Telemetry;
+using VirtualClient.Contracts;
+using VirtualClient.Contracts.Metadata;
+
+namespace VirtualClient.Actions.Kafka
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/Kafka/KafkaClientExecutor.cs:18]: For namespace we can just to VirtualClient.Actions. Apply thoughout

**Review by yangpanMS — COMMENTED** [review]
Would you be ok if you move this PR to the ADO so that we can vet it in QoS_VC internally first?

File: src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs (hunk 15 above)
Comment (mohitgole) [reply src/VirtualClient/VirtualClient.Actions/Kafka/KafkaServerExecutor.cs:358]: Current PR targets support for windows. Linux will be tested and raised as separate PR.

**Comment by mohitgole** [comment]
> Would you be ok if you move this PR to the ADO so that we can vet it in QoS_VC internally first?

Created another ADO PR.
